### ISSUE TITLE
eclipse divergent and other possibilities

### ISSUE THREAD

**Issue by nikosarcevic** [1313e/CMasher#35]
**Describe the colors you would like**
yellow and dark blue/green (like eclipse), orange and gray, orange and purple/or dark brown/or dark blue(blue might be the best)

**Describe the colormap type**
diverging, sequential

**Describe any extra properties**
Colorblind-friendliness; lightness range; color focus; etc.

**Is your colormap request related to a specific application? Please describe.**
I am currently making a lot of plots for different cosmological scenarios. All the scenarios I am inspecting have values above and below a standard one. Therefore a divergent map in some other colors would be great. When I write my thesis eventually I would like to have only one color scheme throughout (that is my dream at least). So far I really like eclipse but it only comes as a sequential. Therefore a divergent eclipse would be grand and you would have my eternal gratitude. Also maybe some pastel options for the cmaps would be great, as someone else requested.

**Additional context**
I posted two plots I've been working on. I used wildfire clipped and not clipped but the color scheme is not to my liking. I think eclipse would be just gorgeous Also thanks for making CMasher - awesome!
<img width="822" alt="Screenshot 2021-04-22 at 10 40 14" src="https://user-images.githubusercontent.com/32773944/115692929-56486780-a357-11eb-86e2-9d8b740aa2e7.png">
<img width="811" alt="Screenshot 2021-04-22 at 10 39 40" src="https://user-images.githubusercontent.com/32773944/115692937-58aac180-a357-11eb-800e-aeaff1c2b2ac.png">

**Comment by 1313e** [comment]
Hi @nikosarcevic,
great to see that you like CMasher.

I can certainly make a diverging colormap that uses `eclipse` on one side.
What color would you like on the other side?
I see you mentioned a few, but that is quite a lot to try out.
I would suggest something like purple or blue, as that is far away from yellow and thus makes it easy to distinguish both sides.

**Comment by nikosarcevic** [comment]
Hi, @1313e ,

thanks for such a prompt response! Yes I agree something like blue would be grand! 
I am not sure how complicated it is to try two options (I assume it is a lot of work . which is why I did not want to generate anything myself as I am still a noob). I still vote for yellow from the eclipse and dark blue (to keep it moody just like eclipse).
Omg if you would do this for me that would be so so nice! My best friend also wants a diverging option of the eclipse!
Keep me posted if you can I will be on the lookout for sure.
Al the best from the UK!

**Comment by nikosarcevic** [comment]
OMG I just saw you are a phd at Swinburne! BTW my best friend is Robert Dzudzar and you know each other! WOW world is so small! Thanks so much for making CMasher it is so great!

**Comment by 1313e** [comment]
> OMG I just saw you are a phd at Swinburne! BTW my best friend is Robert Dzudzar and you know each other! WOW world is so small! Thanks so much for making CMasher it is so great!

Yeah, I'm good friends with Robert as well.
Did he not mention that when telling you to open an issue?

**Comment by nikosarcevic** [comment]
I've discovered CMasher Independently from Robert. Then I mentioned it and it seems he did tell me he knows you but I somehow missed that in our chat some time ago. Then this morning when I realized where you were, I mentioned and he was like "but I told you I know him". LOL me and my brain.
Oh and BTW I saw you are from NL - love the country. Cant wait to travel again so I can go back. (been there many times as my partner is Dutch)

**Comment by 1313e** [comment]
Btw, you could have known that I know Robert if you had read the CMasher paper. ^^

**Comment by nikosarcevic** [comment]
welp! 😂

**Comment by 1313e** [comment]
Anyway, combining two existing colormaps with each other to make a diverging colormap does not take that long.
Give me a minute.

**Comment by 1313e** [comment]
This is `cosmic` plus `eclipse` (and a few modifications to make it look correct):
![cosmic_eclipse](https://user-images.githubusercontent.com/30159027/115802789-df78a080-a422-11eb-9370-e0c93c5db163.png)

**Comment by nikosarcevic** [comment]
looks nice. sorry for nagging but is it possible to just take the eclipse seq and make it div? does it make any sense in terms of color theory? because I really really like just the colors in the eclipse

(side comment unrelated: when I submitted an issue, ive noticed my comments were not there. I had to remove the HTML part of the code in the template  (where you place the text, meaning the <> and then my comments turned up. I guess that was the problem with the other issue where someone asked for pastel colors)

**Comment by 1313e** [comment]
How would a diverging colormap work that has the same colors on both sides?
Then you cannot tell the difference between left and right.

**Comment by 1313e** [comment]
Do you want me to provide a Python standalone version of the colormap above, such that you can try it out in your plots and then give me feedback on it?

**Comment by nikosarcevic** [comment]
Hi @1313e 
sorry for not responding sooner, I was offline for the weekend.

I didn't express myself well in the previous part - there are basically 3 main colors in the eclipse: yellowish, greenish and black. so if we can have yellow on one side, that teal/green/blue on the other and then black in the middle? or something like that? actually not sure if it would make sense?

Second Q you asked: yes, sure I can try! I need to do more plotting today so it would be perfect

**Comment by 1313e** [comment]
> I didn't express myself well in the previous part - there are basically 3 main colors in the eclipse: yellowish, greenish and black. so if we can have yellow on one side, that teal/green/blue on the other and then black in the middle? or something like that? actually not sure if it would make sense?

This could be done, but then I need to come up with colors that fill up the remaining spaces.
Because, pretty much, that would result in the yellow side requiring filler darker colors and the green side requiring filler brighter colors.
After all, `eclipse` is a sequential colormap, and diverging colormaps are pretty much two sequential colormaps slammed together, so pretty much you are asking me to split a single sequential colormap up into two of them.

It might be possible, but it is incredibly hard to do.
Especially as I cannot guarantee in any way that the colors of `eclipse` will even be in such a diverging colormap.
I might be able to get close, but get it all the way is in no way guaranteed.

Also, even if I were to do so, the resulting colormap will be hard to read for anyone with colorblindness.
I am assuming you want to avoid that.


> Second Q you asked: yes, sure I can try! I need to do more plotting today so it would be perfect

Here you go: [celestial.zip](https://github.com/1313e/CMasher/files/6374997/celestial.zip)
It contains a stand-alone Python module of the colormap.
Simply put this file in your working directory and import it with `import celestial` (`celestial` is the current working name for the colormap).
This will register it in MPL such that you can access it the way you can with any CMasher colormaps.
Let me know what you think about it and whether you want changes to it, or something completely different.

PS: I personally really like the combination of `eclipse` and `cosmic` into `celestial`, so even if you don't like it, I will probably end up adding it to CMasher.
Maybe after some small adjustments though.

**Comment by nikosarcevic** [comment]
thank you so much. I agree with your points, you have much more experience in this than I do (color theory). I need to finish up some administrative stuff quickly but I can share the results (plots) and my thoughts here soon.

**Comment by nikosarcevic** [comment]
Hello, @1313e 

as you can imagine, the admin stuff has gotten out hand and had teaching to do yesterday so didn't manage to send you the update - but I did test it. Here is the plot in celestial: looks good, cannot argue. Now I do not want to sound horrible or ungrateful but I do not think (damn this is so hard to say) I will be using it for these plots. I still would like some other combination.
![pk_lin_omega_c](https://user-images.githubusercontent.com/32773944/116192107-4c42b200-a725-11eb-9ede-299e1cea3b22.png)

**Comment by 1313e** [comment]
> Hello, @1313e
> 
> as you can imagine, the admin stuff has gotten out hand and had teaching to do yesterday so didn't manage to send you the update - but I did test it. Here is the plot in celestial: looks good, cannot argue. Now I do not want to sound horrible or ungrateful but I do not think (damn this is so hard to say) I will be using it for these plots. I still would like some other combination.

That is fine, don't worry about it.
Do you know any colors you would like to replace the blue side with?
It is probably better if you could describe it in terms of colormaps that CMasher already has, as that gives me (and you) a much better idea what surrounding colors would likely end up on that side.

**Comment by nikosarcevic** [comment]
hmm something like sepia, arctic or fall?

**Comment by 1313e** [comment]
> hmm something like sepia, arctic or fall?

Let me have a look at that.

**Comment by nikosarcevic** [comment]
ok I need to go do some quick chores and shower so ill be away from my laptop for an hour or so (in case you ask me something)

**Comment by 1313e** [comment]
Alright, I have 4 concept colormaps for you:
![arctic_eclipse](https://user-images.githubusercontent.com/30159027/116337520-960ec380-a81d-11eb-90cb-9ee52947691e.png)
![arctic_eclipse_extended](https://user-images.githubusercontent.com/30159027/116337524-9909b400-a81d-11eb-8a67-784ef9f1577a.png)
![fall_eclipse](https://user-images.githubusercontent.com/30159027/116337527-99a24a80-a81d-11eb-9623-981da1244739.png)
![sepia_eclipse](https://user-images.githubusercontent.com/30159027/116337530-99a24a80-a81d-11eb-8ecc-5baa4494f0de.png)

The non-extended ones is where i literally took both colormaps and combined them together with the minimal amount of changes to either side to get it to be perceptually uniform sequential.
The `arctic_eclipse_extended`, I extended the `arctic` side such that the full `eclipse` side can be included, which meant introducing a lot more blue and purple to the colormap.
It actually now comes kinda close to the `celestial` colormap I showed before.
The `fall_eclipse` and `sepia_eclipse` colormaps are actually much more like cyclic colormaps than diverging colormaps due to them having the same ending color.
I do actually like both of them, so maybe I will add them in some capacity.

In case you want to try out any of these, I attach them in this zipped file: 
[colormaps.zip](https://github.com/1313e/CMasher/files/6388717/colormaps.zip)

**Comment by nikosarcevic** [comment]
W O W

this is amazing. this is awesome. wow wow wow!!

thank you so so much!!!!!

**Comment by 1313e** [comment]
@nikosarcevic Did you try out the new colormaps? :)

**Comment by nikosarcevic** [comment]
Hi, @1313e 
YES! Sorry forgot to share here. Loving it! Thanks so much. Are you planning on implementing it into cmr soon? Thanks a bunch again :much obliged:
<img width="851" alt="Screenshot 2021-05-03 at 07 21 24" src="https://user-images.githubusercontent.com/32773944/116846540-60ccf180-abe0-11eb-8396-49db1f321f12.png">

**Comment by 1313e** [comment]
I am definitely not adding all of them.
I am just asking you which one you like the most and whether I should make more changes somewhere.
As far as I can see, you are using the `sepia_eclipse` colormap, which is more a cyclic colormap and not a diverging colormap, as you can see in the colorspace overview.
So, if I would add it, I would make several changes to it to make it a diverging colormap.

**Comment by nikosarcevic** [comment]
I think I am using a fall_eclipse one with trimmed "edges" in that one. I like the last two ones the most (sepia_eclipse and fall_eclipse). Would those two be too much to ask? If I have to choose only one it would be fall_eclipse I think. :)

**Comment by 1313e** [comment]
> I think I am using a fall_eclipse one with trimmed "edges" in that one. I like the last two ones the most (sepia_eclipse and fall_eclipse). Would those two be too much to ask? If I have to choose only one it would be fall_eclipse I think. :)

That one is also cyclic actually.
So, I will have to trim it.
I will have a look in a few hours to see what I can do about it.

**Comment by nikosarcevic** [comment]
no rush from my side, I have your code :)
(although I have noticed it slows down my computing A LOT. I just jupyter nb for plotting as it is the easiest. so I just copy your code into a cell, rename the map and use it that way. it really makes to nb run super slow. not sure if that is the best way to use it - am I doing it a noob way? XD)

**Comment by 1313e** [comment]
Why don't you just import it?
Just put the Python file in the same directory as where your Jupyter notebook is, and then just import the colormap with `import <colormap_name>`.
After that, you can use it the same way you could normally do through MPL.
However, it can take a little while to import it the first time, as creating and registering a colormap takes a little while.

**Comment by nikosarcevic** [comment]
ah ok I will try it that way! thanks and sorry for being a noob 😂

**Comment by 1313e** [comment]
Actually, I could add either of them to CMasher as a cyclic colormap, and one can convert it to a diverging colormap by cutting off the edges using CMasher.

Anyway, you can of course pick all of them.
However, I won't add all of them to CMasher. 
If you want all of them, I will give you the source files as well after cleaning them up a bit.

**Comment by nikosarcevic** [comment]
No problem and yes please share the code once you do your magic! All the best from NCL!

**Comment by 1313e** [comment]
Alright, let me perform some arcane rituals...., I mean, magic...

**Comment by 1313e** [comment]
Alright, so I took the `fall_eclipse` and `sepia_eclipse` colormaps I showed above and turned them into actual proper colormaps.
Below are their `viscm` outputs:
![fall_eclipse_cyclic](https://user-images.githubusercontent.com/30159027/116959134-031fce80-ace0-11eb-84c7-bbedd19d9807.png)
![sepia_eclipse_cyclic](https://user-images.githubusercontent.com/30159027/116959139-0450fb80-ace0-11eb-80d2-c44458f7fd8a.png)
As you can see, both of them are cyclic colormaps now, not diverging colormaps.
Also, I had to cut a bit of yellow off the `fall_eclipse` one, as otherwise the ends would become incredibly hard to distinguish due to them overlapping a lot.
Both of these colormaps are finished (well, besides naming them), so below you can download both of the colormaps:
[colormaps.zip](https://github.com/1313e/CMasher/files/6418922/colormaps.zip)

Do remember that these are cyclic colormaps and thus upon importing either of them (use `import <cmap_name>` for that), they will be registered as such.
This means that the normal version, the shifted version (which has dark in the center instead of bright), and the reversed versions of both will be registered.
If I am not mistaken, you want a dark center, so you will have to use the ones that have a `_s` prefix in MPL.
Also, because they are cyclic, their ends are the same, so you will have to cut them off is you want to use them as diverging colormaps.

As that might be a bit hard to figure out, here is how you would (probably) use the `fall_eclipse_cyclic` colormap from above as a diverging colormap with a black center for your plot:
```python
import fall_eclipse_cyclic   # Import colormap from the colormap module into MPL
import cmasher as cmr

# Create a sub-colormap of 'cmr.fall_eclipse_cyclic_s' to use as a diverging colormap
# Either use [0.05, 0.95] as I do below, or [0.1, 0.9] if the ends are still too similar
cmap = cmr.get_sub_cmap('cmr.fall_eclipse_cyclic_s', start=0.05, stop=0.95)

# Now whenever you want to use the colormap, simply pass the `cmap` object to the function
# instead of the name of the colormap (which is what I assume you usually do)
```

Let me know if you are satisfied with these colormaps.
Personally, I like both quite a lot and might end up adding them both actually.
In that case, they will have proper names of course.

**Comment by nikosarcevic** [comment]
That looks absolutely stunning!!! WOW
You , sir, [are a wizard](http://gph.is/2noyXQD)! 

I will have to try to see if I have the time to test today - I am teaching two courses and attending a conference and have some work to finish today but I will report back as soon as I can.
Thank yo use much for this.

**Comment by 1313e** [comment]
_Patiently waiting_

**Comment by nikosarcevic** [comment]
@1313e I am so sorry! I simply do not have the time this week. I have a proposal to finish, two conferences and respective meetings, teaching and some other projects to wrap up that are due by Monday. I am simply going mad here :cry:

I am almost certain I did test the fall_eclipse one on Monday lemme quickly check

**Comment by 1313e** [comment]
Well, then I will see your comments arrive next week.
No need to hurry.
As I said, I am patiently waiting.

**Comment by nikosarcevic** [comment]
thanks! it seems I did not test it then :(

off I go shower and continue writing the proposal (wish me luck!)

**Comment by nikosarcevic** [comment]
btw are you on twitter? or anywhere on SM I can get in touch other than here?

**Comment by 1313e** [comment]
> btw are you on twitter? or anywhere on SM I can get in touch other than here?

The only place would be Facebook then I guess.

**Comment by nikosarcevic** [comment]
ahahah well there is an update. I was installing some cosmo related libraries, made a mess, nothing works including my ipython 😂😂😭😭😭😭 have to try to reposter everything and install what I need. wish me luck. so stoooopid 😂😂😂

**Comment by nikosarcevic** [comment]
so after taking a substantial amount of time to make everything work (yay me), here are the plots:

<img width="799" alt="Screenshot 2021-05-13 at 22 40 04" src="https://user-images.githubusercontent.com/32773944/118191363-5b6f7280-b43c-11eb-88dd-182bed1963ff.png">
<img width="767" alt="Screenshot 2021-05-13 at 22 40 22" src="https://user-images.githubusercontent.com/32773944/118191369-5d393600-b43c-11eb-8b3c-78826716f589.png">


loving it, cant really decide between the two <3 TYSM

**Comment by nikosarcevic** [comment]
<img width="777" alt="Screenshot 2021-05-13 at 22 50 47" src="https://user-images.githubusercontent.com/32773944/118192198-ae95f500-b43d-11eb-9d1c-c2fca2d4eb20.png">

and the cyclic version :)

**Comment by 1313e** [comment]
You approve of both colormaps? 😄

**Comment by nikosarcevic** [comment]
> You approve of both colormaps? 😄
> Guess I have to get to naming them now, huh?

here is an official niko blessing :waves hands in the blessing way, chanting: 😂😂😂

lemme know if you need help naming.

**Comment by 1313e** [comment]
If you have any suggestions, I will take them.

**Comment by nikosarcevic** [comment]
fall eclipse cyclic looks like it could be "amber" and sepia eclipse cyclic looks like "haze" in my brain 😎

**Comment by 1313e** [comment]
Well, `amber` is already a colormap, so that is kinda hard, but `haze` sounds pretty cool actually.

**Comment by nikosarcevic** [comment]
ah damn I mena ember but it corrected me to amber. but yeah that would be too ambiguous. lemme think.

**Comment by 1313e** [comment]
`ember` also already exists. :laughing:

**Comment by nikosarcevic** [comment]
r u serious? why didn't I look it up before!!!!!and I use cmasher hahahahaha :what a facepalm.

I need to wake up properly lemme make some coffee 😂😂😂

**Comment by 1313e** [comment]
If you have CMasher v1.6.1 installed, you can get a list of all names by executing `cmr cmlist` on the commandline.

**Comment by nikosarcevic** [comment]
what about tinder? it reminds me of something fire related?

**Comment by 1313e** [comment]
I will take them into consideration. :)

**Comment by 1313e** [comment]
So, I have decided to not add the `fall_eclipse_cyclic` to CMasher, as it is simply too similar to `pride`.
Therefore, it will also not receive a name.
Which means, you can give it whatever name you feel like (just modify it both within the file and the name of the file).

The `sepia_eclipse_cyclic` will receive a name fairly soon.

**Comment by nikosarcevic** [comment]
> So, I have decided to not add the `fall_eclipse_cyclic` to CMasher, as it is simply too similar to `pride`.
> Therefore, it will also not receive a name.
> Which means, you can give it whatever name you feel like (just modify it both within the file and the name of the file).
> 
> The `sepia_eclipse_cyclic` will receive a name fairly soon.

no problemo! I have "my own" ahahaha. Thanks so much for this! BTW Robert texted me saying "57 messages on GH issues? you guys are insane"

**Comment by 1313e** [comment]
Alright, `sepia_eclipse_cyclic` was added as `copper` in v1.6.2, so I will mark this issue as complete.

PS: I also added the colormap `ghostlight`, which you might be interested in actually.

**Comment by nikosarcevic** [comment]
Copper is such a grand name! Everything looks so good, thanks so so much.

BTW I noticed prinsenvlag is a new one  and I love it! Name is perfect (prima!)
